lib/mpool/include/mpool/mpool.h

```diff
@@ -441,6 +441,20 @@ mpool_mblock_write(struct mpool *mp, uint64_t mbid, const struct iovec *iov, int
 merr_t
 mpool_mblock_read(struct mpool *mp, uint64_t mbid, const struct iovec *iov, int iovc, off_t offset);
 
+/**
+ * mpool_mblock_clone() - clone the specified mblock
+ *
+ * @mp:       mpool
+ * @mbid:     mblock object ID to clone from
+ * @off:      start offset to clone from/to in the source/target mblock id
+ * @len:      number of bytes to clone. SIZE_MAX implies the entire mblock.
+ * @mbid_out: target mblock id (output)
+ *
+ * Return: %0 on success, <%0 on error
+ */
+merr_t
+mpool_mblock_clone(struct mpool *mp, uint64_t mbid, off_t off, size_t len, uint64_t *mbid_out);
```

Comment:
Function header should describe in more detail what it means to use `off > 0` or `len < old.write_len`.  For example: creates a sparse file, reading "the holes" will return 0s, and does it depend on the filesystem?

Also document allocated and write len of new mblock.  For example:
- `new.write_len == off + len`
- `new.allocated_len >= new.write_len`  (which is probably less specific than it has to be, but that leaves room for change).

Also, would it be insane to let `len==0` imply full clone?  Or `len > old.write_len`?  Or `len == (size_t)-1`?  I hate having to write `SIZE_MAX` in the default case.

Reply:
Thanks for the detailed feedback! Have changed the function header to describe the requirements in detail.

I don't think using len == 0 to imply a full clone is insane. That's where I started but ended up using SIZE_MAX to mimic how other APIs, like mpool_mcache_madvise(), behave. Now that I have a second vote for "len==0", let me change it.

Reply:
Re: whether returning zeroes depend on the filesystem:

All file-systems that support PUNCH_HOLE must return zeroes when reading back from a punched region. We document that HSE works only on file systems that supports this operation.

Reply:
I have seen many APIs that use -1 to mean copy the whole thing. There are glib string APIs like that.

> All file-systems that support PUNCH_HOLE must return zeroes when reading back from a punched region. We document that HSE works only on file systems that supports this operation.

mblock_clone() wouldn't work on non-PUNCH_HOLE file systems?

Reply:
HSE relies on the file system to support fallocate with PUNCH_HOLE. This is used by mpool to discard file-system blocks during mblock deletion. mblock_clone just rides on it, so if cN attempts to read a region from the target mblock that's not cloned from the source mblock it gets zeroes.

Reply:
There's some precedence in filesystem APIs (posix_fadvise(2), sync_file_range(2) etc.) where len == 0 implies EOF. Would like to stick with len == 0 to imply a full clone.